### PR TITLE
Performance optimization on speed and memory for RDD

### PR DESCRIPTION
**Pull request by asaidesita**
This pull request is similar to #59.
This implementation manages caching and evicting RDDs on GPU automatically
in order to redundant data transfers between a host PC and the GPU.
There are two differences between #59 and this.

## Difference:
  1. Optimize data transfers between CPU and GPU between multiple GPUEnabler _RDD_ APIs calls , not for _DataSet_ APIs.
  2. Handle auto-cached RDDs on GPU cleanup based on a LRU policy
      in the once execution.
      Manage only auto-cached RDDs on GPU automatically.
      If GPU free space becomes small less than a certain value and 
      GPUEnabler can't transfer a new RDD from a host PC to the GPU, 
      this implementation evicts old auto-cached RDDs based on a LRU policy automatically.

**Note**: Adoption of LRU creates the possibility of reusing cached RDDs on GPU over multiple jobs.

### CODE REVIEW

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/ColumnPartitionSchema.scala [hunk 1]
```diff
@@ -36,7 +44,7 @@ private[gpuenabler] case object ColumnFormat extends PartitionFormat
 
 private[gpuenabler] object ColumnPartitionSchema {
 
-  // Since we are creating a runtime mirror using the class loader of current thread,
+  // Since we are creating a runtime mirror usign the class loader of current thread,
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/ColumnPartitionSchema.scala]: nit: using?

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/GPUMemoryManager.scala [hunk 2]
```diff
@@ -121,20 +158,56 @@ private[gpuenabler] class GPUMemoryManager(val executorId : String,
...
     }
   }
+  def autoUnCacheGPU(rddId : Int): Unit = {
+    autoCachedGPURDDs.foreach(s =>
+      if(cachedGPURDDs.contains(s)) {
+          autoCachedGPURDDs -= s
+      }
+    )
+    if(autoCachedGPURDDs.size > 0){
+     val rddId = autoCachedGPURDDs.head
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/GPUMemoryManager.scala]: nit: Can we format this block with better indent?

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/ColumnPartitionSchema.scala [hunk 3]
```diff
@@ -252,15 +260,15 @@ private[gpuenabler] class ColumnSchema(
...
+  private def readObject(in: ObjectInputStream): Unit = CUDAUtils.sparkUtils.tryOrIOException { 
     val mirror = ColumnPartitionSchema.mirror
     _columnType = in.readObject().asInstanceOf[ColumnType]
     _terms =
       in.readObject().asInstanceOf[Vector[(String, String)]].map { case (clsName, propName) =>
         val cls = CUDAUtils.sparkUtils.classForName(clsName)
         val typeSig = mirror.classSymbol(cls).typeSignature
-        typeSig.decl(universe.TermName(propName)).asTerm
-        // typeSig.declaration(universe.stringToTermName(propName)).asTerm // Scala_2.10
+        typeSig.declaration(universe.stringToTermName(propName)).asTerm
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/ColumnPartitionSchema.scala]: Can we use scala 2.11 as default?

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala [hunk 4]
```diff
@@ -316,7 +403,6 @@ object CUDARDDImplicits {
                       outputArraySizes: Seq[Int] = null,
                       inputFreeVariables: Seq[Any] = null): T = {
       import org.apache.spark.gpuenabler.CUDAUtils
-
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: Can we revert this since this change is not necessary?

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala [hunk 5]
```diff
@@ -352,7 +438,6 @@ object CUDARDDImplicits {
           }
         }
       }
-
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala:355]: Can we revert this since this change is not necessary?

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala [hunk 6]
```diff
@@ -237,7 +240,85 @@ object CUDARDDImplicits {
...
+     * Classified number 1 and 0 is a dependency type deviding a stage
+     * or not respectively.
+     */
 
+      def judgeApiType(rdd: RDD[_]): Int = {
+        val rddType = rdd.getClass()
+        val mapGPUPartitionsClass = classOf[MapGPUPartitionsRDD[_,_]]
+        val convertGPUPartitionsClass = classOf[ConvertGPUPartitionsRDD[_]]
+        rddType match {
+          case x if x==mapGPUPartitionsClass => 1
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: nit: Can we add ` ` before and after `==`?

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala [hunk 7]
```diff
@@ -237,7 +240,85 @@ object CUDARDDImplicits {
...
+              classifyDep(d.rdd, apiType, gpuApiFlag)
+            })
+            val lastDep = rdd.dependencies.last
+            val apiType = judgeApiType(lastDep.rdd)
+            classifyDep(lastDep.rdd, apiType, gpuApiFlag)
+        }
+      }
+
+      def classifyDep(rdd: RDD[_], apiType: Int, gpuApiFlagArg: Int): Unit = {
+        var gpuApiFlag = gpuApiFlagArg
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: Is this variable necessary? Only at line 289, `gpuApiFlag` is read. In other places, `gpuApiFlag` may not be required or can be used as `val`.

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala [hunk 8]
```diff
@@ -141,7 +142,7 @@ private[gpuenabler] class ConvertGPUPartitionsRDD[T: ClassTag](
         hyIter
       }
       case iter: Iterator[T] => {
-        // println("Converting Regular Iterator to hybridIterator")
+    	// println("Converting Regular Iterator to hybridIterator")
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala:145]: nit: can we revert this change?

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala [hunk 9]
```diff
@@ -115,7 +116,7 @@ private[gpuenabler] class MapGPUPartitionsRDD[U: ClassTag, T: ClassTag](
       f(context, split.index, firstParent[T].iterator(split, context))
     }
   }
-}
+}
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: nit: can we revert this change?

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDAManager.scala [hunk 10]
```diff
@@ -132,6 +132,31 @@ private[gpuenabler] class CUDAManager {
...
+     * this method is called on allocateGPUMemory.
+     * In order to reuse cached RDDs on gpu as much as possible,
+     * adopt LRU as a cache strategy.
+     *
+     * gpuMemThreshold is a threshold for sadety of a GPU memory.
+     * when this variable has "1000000000",
+     * if GPU free space become less than "1000000000B (1 GB)",
+     * auto cached RDD on GPU is evicted on a LRU policy.
+     */
+    val gpuMemThreshold = 1000000000 //1GB
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDAManager.scala]: Can we make this value `1000000000` configurable? For example, read this value from a property or option.

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala [hunk 11]
```diff
@@ -237,7 +240,85 @@ object CUDARDDImplicits {
...
+        * every time GPUEnabler invoke actions has a sequent of GPU methods,
+        *
+        * gpuApiFlag shows the child type of a RDD.
+        * When the child RDD's API
+        *  case 0 => CPU API (ex, mapRDD)
+        *  case 1 => GPU API (don't divide a stage. ex, mapGPURDD)
+        *  case 2 => GPU API (divide a stage into two. ex,shuffleGPURDD)
+        */
+      var gpuApiFlag = 0
+      if(judgeApiType(rdd) == 1) rdd.autoCacheGpu(); gpuApiFlag = 1
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: Is this line correct? The current version always set `1` into `gpuApiFlag`.
If this is correct, we can replace line 258 with `val gpuApiFlag = 1`.

Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: May I ask you whether this line is correct? The current version always sets 1 into gpuApiFlag.
If this is correct, we can replace line 258 with val gpuApiFlag = 1.

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDAManager.scala [hunk 12]
```diff
@@ -132,6 +132,31 @@ private[gpuenabler] class CUDAManager {
...
+    /* remove cached RDDs on GPU automatically
+     * without ignoring user's cache (unCache) indication.
+     * Since spark can't estimate RDD size via distributed data
+     * on eacn nodes, this method removes cached RDDs on GPU
+     * in the partitions units.
+     * this method is called on allocateGPUMemory.
+     * In order to reuse cached RDDs on gpu as much as possible,
+     * adopt LRU as a cache strategy.
+     *
+     * gpuMemThreshold is a threshold for sadety of a GPU memory.
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDAManager.scala]: nit: `safety`?

**Comment by asaidesita** [comment]
Can I correct and repost this pull request?
I am sorry for you that there are too much mistakes.
（If the way you revise are easier and quicker for you to handle, I agree your revisions.)

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/ColumnPartitionSchema.scala (hunk 1 above)
Comment (asaidesita) [reply gpu-enabler/src/main/scala/com/ibm/gpuenabler/ColumnPartitionSchema.scala]: You are correct.
correct : using
error : usign

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/GPUMemoryManager.scala (hunk 2 above)
Comment (asaidesita) [reply gpu-enabler/src/main/scala/com/ibm/gpuenabler/GPUMemoryManager.scala]: Yes you can. 
This indent has no intention.

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/ColumnPartitionSchema.scala (hunk 3 above)
Comment (asaidesita) [reply gpu-enabler/src/main/scala/com/ibm/gpuenabler/ColumnPartitionSchema.scala]: No.
I want to revise this lines and use scala 2.10 as default.
I didn't confirm the whole code in order to modify scala version.

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala (hunk 4 above)
Comment (asaidesita) [reply gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: Yes. No ploblem.
Sorry to make trivial mistakes.

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala (hunk 5 above)
Comment (asaidesita) [reply gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala:355]: Yes. No ploblem.
Sorry to make trivial mistakes.

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala (hunk 6 above)
Comment (asaidesita) [reply gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: Yes.
This expression has no intention.

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala (hunk 7 above)
Comment (asaidesita) [reply gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: This variable is not necessary.
I want to delete line 283 and revise line 289.

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala (hunk 8 above)
Comment (asaidesita) [reply gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala:145]: Yes.I mistakenly touched it.

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala (hunk 9 above)
Comment (asaidesita) [reply gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: Yes.I mistakenly touched it.

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDAManager.scala (hunk 10 above)
Comment (asaidesita) [reply gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDAManager.scala]: Yes you can.
If necessary, I will add that property (or option).

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala (hunk 11 above)
Comment (asaidesita) [reply gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: This line is incorrect.
Because I assume that autoCacheGPU method is called in an action using GPUs,
it is correct for this flag to have "1".

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDAManager.scala (hunk 12 above)
Comment (asaidesita) [reply gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDAManager.scala]: Yes.
correct : safety
error    : sadety

**Comment by kiszk** [comment]
Could you update this pull request by adding other commits based on discussions? It would be good to have some iterations for a pull request.
After some iterations, I hope we can merge it into master.

**Comment by asaidesita** [comment]
Yes, I will.
I update this pull request as you said.
Thank you for your attention to this matter.

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/GPUMemoryManager.scala [hunk 13]
```diff
@@ -121,20 +158,56 @@ private[gpuenabler] class GPUMemoryManager(val executorId : String,
...
   }
+  def autoUnCacheGPU(rddId : Int): Unit = {
+    autoCachedGPURDDs.foreach(s =>
+      if(cachedGPURDDs.contains(s)) {
+          autoCachedGPURDDs -= s
+      }
+    )
+    if(autoCachedGPURDDs.size > 0){
+      val rddId = autoCachedGPURDDs.head
+    autoCachedGPURDDs -= rddId
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/GPUMemoryManager.scala]: Is it possible to improve the indentation?

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/ColumnPartitionSchema.scala [hunk 14]
```diff
@@ -21,12 +21,20 @@ import java.io.{ObjectInputStream, ObjectOutputStream}
 
 import org.apache.spark.gpuenabler.CUDAUtils
 import org.apache.spark.util.Utils
+import org.apache.spark.gpuenabler.CUDAUtils.sparkUtils.tryOrIOException
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/ColumnPartitionSchema.scala]: While I am not sure whether all of new imports are necessary, could you double-check whether they are necessary?
(e.g. `scala.reflect.ClassTag` is imported twice).

Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/ColumnPartitionSchema.scala]: While I am not sure whether all of new imports are necessary, could you double-check whether they are necessary?
(e.g. scala.reflect.ClassTag is imported twice).

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/ColumnPartitionSchema.scala [hunk 15]
```diff
@@ -252,7 +260,7 @@ private[gpuenabler] class ColumnSchema(
     out.writeObject(propertyChain)
   }
 
-  private def readObject(in: ObjectInputStream): Unit = CUDAUtils.sparkUtils.tryOrIOException {
+  private def readObject(in: ObjectInputStream): Unit = CUDAUtils.sparkUtils.tryOrIOException {
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/ColumnPartitionSchema.scala]: nit: Would it be possible to revert this change to drop ` ` at the end of the line?

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala [hunk 16]
```diff
@@ -237,7 +240,78 @@ object CUDARDDImplicits {
...
+              val apiType = judgeApiType(d.rdd)
+              classifyDep(d.rdd, apiType, gpuApiFlag)
+            })
+            val lastDep = rdd.dependencies.last
+            val apiType = judgeApiType(lastDep.rdd)
+            classifyDep(lastDep.rdd, apiType, gpuApiFlag)
+        }
+      }
+
+      def classifyDep(rdd: RDD[_], apiType: Int, gpuApiFlag: Int): Unit = {
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: Would it be possible to define some constant literals for values of `gpuApiFlag` (i.e. `0`, `1`, `2`)?
It is ease to understand source code.

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala [hunk 17]
```diff
@@ -237,7 +240,78 @@ object CUDARDDImplicits {
...
+     * Classified number 1 and 0 is a dependency type deviding a stage
+     * or not respectively.
+     */
 
+      def judgeApiType(rdd: RDD[_]): Int = {
+        val rddType = rdd.getClass()
+        val mapGPUPartitionsClass = classOf[MapGPUPartitionsRDD[_,_]]
+        val convertGPUPartitionsClass = classOf[ConvertGPUPartitionsRDD[_]]
+        rddType match {
+          case x if x == mapGPUPartitionsClass => 1
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: ditto

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala (hunk 8 above)
Comment (kiszk) [reply gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala:145]: nit: Would it be possible to revert this change?

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala [hunk 18]
```diff
@@ -237,7 +240,78 @@ object CUDARDDImplicits {
...
+          case x if x == convertGPUPartitionsClass=> 1
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: Would it be possible to integrate two `case x ...` into one `case x ...` using `case x if ((x == ...) || (x = ...)) ...` ?

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala [hunk 19]
```diff
@@ -237,7 +240,78 @@ object CUDARDDImplicits {
...
+            if(gpuApiFlag == 1) rdd.autoCacheGpu()
+            trackLineage(rdd, 1)
+          case 2 =>
+            trackLineage(rdd, 1)
+        }
+      }
+
+   /**
+     * This function judges whether a RDD's dependency type is
+     * to devide a stage or not.
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: nit: `divide`?

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala [hunk 20]
```diff
@@ -237,7 +240,78 @@ object CUDARDDImplicits {
...
+     * when implement a new GPUPartitionsRDD,
+     * to classify and write a new it on this method's "match" branch avoids
+     * redundant data transfers between GPU APIs.
+     * Classified number 1 and 0 is a dependency type deviding a stage
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/CUDARDDUtils.scala]: nit: `divide`?

**Comment by kiszk** [comment]
@asaidesita would it be possible to resolve conflicts, too?
We expect that we expect to merge this soon (one or two more iterations)

**Comment by asaidesita** [comment]
Yes, sure.
I'm sorry that I am late.
I will solve this problem in a week

File: gpu-enabler/src/main/scala/com/ibm/gpuenabler/GPUSparkEnv.scala [hunk 21]
```diff
@@ -80,6 +80,39 @@ private[gpuenabler] object GPUSparkEnv {
...
+    if (SparkEnv.get.conf.getInt("spark.gpuenabler.autocacheevict.isenabled", 1) == 1) true else false
+
+  val gpuMemCap: Long= {
+    import jcuda.runtime.JCuda
+    val freeMemArray = new Array[Long](1)
+    val totalMemArray = new Array[Long](1)
+    jcuda.runtime.JCuda.cudaMemGetInfo(freeMemArray, totalMemArray)
+    val evictThreshold = SparkEnv.get.conf.getInt("spark.gpuenabler.autocacheevict.gpumem", 0)
+    if (isAutoCacheEvictEnabled) {
+      if (evictThreshold <= totalMemArray(0) && evictThreshold > 0){
```
Comment (kiszk) [inline gpu-enabler/src/main/scala/com/ibm/gpuenabler/GPUSparkEnv.scala]: I think that it is good to introduce these values in `GPUSparkEnv`.

Since `totalMemArray(0)` is [represented `in bytes`](http://www.jcuda.org/jcuda/doc/jcuda/runtime/JCuda.html#cudaMemGetInfo(long[],%20long[])), do we need to perform a multiplication for `evictThreshold` before this `if` statement?

**Comment by kiszk** [comment]
Sorry for being late to review this. LGTM.
In addition, it would be good to see this paper https://dl.acm.org/citation.cfm?id=3180276 written by a contributor.

**Comment by kiszk** [comment]
Thank for this work. Merged.

**Comment by asaidesita** [comment]
Thank you for your accurate and sincere reviews despite your busy schedule.
I am honored to hear your comment for our paper.  
I had a valuable experience thanks to you and your drastic product IBMSparkGPU.
